mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/task/MailboxMergingTaskRunner.java

```diff
@@ -0,0 +1,94 @@
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.cassandra.mail.task;
+
+import javax.inject.Inject;
+
+import org.apache.james.mailbox.MailboxManager;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.acl.ACLDiff;
+import org.apache.james.mailbox.cassandra.ids.CassandraId;
+import org.apache.james.mailbox.cassandra.mail.CassandraACLMapper;
+import org.apache.james.mailbox.cassandra.mail.CassandraMailboxDAO;
+import org.apache.james.mailbox.cassandra.mail.CassandraMessageIdDAO;
+import org.apache.james.mailbox.cassandra.mail.CassandraUserMailboxRightsDAO;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.ComposedMessageId;
+import org.apache.james.mailbox.model.ComposedMessageIdWithMetaData;
+import org.apache.james.mailbox.model.MailboxACL;
+import org.apache.james.mailbox.model.MessageRange;
+import org.apache.james.mailbox.store.StoreMessageIdManagerNoChecks;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MailboxMergingTaskRunner {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MailboxMergingTaskRunner.class);
+
+    private final MailboxManager mailboxManager;
+    private final StoreMessageIdManagerNoChecks messageIdManager;
+    private final CassandraMessageIdDAO cassandraMessageIdDAO;
+    private final CassandraMailboxDAO mailboxDAO;
+    private final CassandraUserMailboxRightsDAO rightsDAO;
+    private final CassandraACLMapper cassandraACLMapper;
+
+    @Inject
+    public MailboxMergingTaskRunner(MailboxManager mailboxManager, StoreMessageIdManagerNoChecks messageIdManager, CassandraMessageIdDAO cassandraMessageIdDAO, CassandraMailboxDAO mailboxDAO, CassandraUserMailboxRightsDAO rightsDAO, CassandraACLMapper cassandraACLMapper) {
+        this.mailboxManager = mailboxManager;
+        this.messageIdManager = messageIdManager;
+        this.cassandraMessageIdDAO = cassandraMessageIdDAO;
+        this.mailboxDAO = mailboxDAO;
+        this.rightsDAO = rightsDAO;
+        this.cassandraACLMapper = cassandraACLMapper;
+    }
+
+    public Task.Result run(CassandraId oldMailboxId, CassandraId newMailboxId) throws MailboxException {
+        MailboxSession session = mailboxManager.createSystemSession("task");
+        return moveMessages(oldMailboxId, newMailboxId, session)
+            .ifCompleted(() -> mailboxDAO.delete(oldMailboxId).join())
+            .ifCompleted(() -> mergeRights(oldMailboxId, newMailboxId));
```

Comment:
Shouldn't you merge right before removing the old mailbox?

Reply:
Maybe for readability but it have no technical impotance.

Reply:
I inversed them for readability.